controllers/reconcilers/grafana/httproute_reconciler.go

```diff
@@ -0,0 +1,208 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+const (
+	protocolHTTP  = "http"
+	protocolHTTPS = "https"
+)
+
+type HTTPRouteReconciler struct {
+	client client.Client
+}
+
+func NewHTTPRouteReconciler(client client.Client) reconcilers.OperatorGrafanaReconciler {
+	return &HTTPRouteReconciler{
+		client: client,
+	}
+}
+
+func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, cr *v1beta1.Grafana, vars *v1beta1.OperatorReconcileVars, scheme *runtime.Scheme) (v1beta1.OperatorStageStatus, error) {
+	log := logf.FromContext(ctx).WithName("HTTPRouteReconciler")
+
+	if cr.Spec.HTTPRoute == nil {
+		return v1beta1.OperatorStageResultSuccess, nil
+	}
+
+	log.Info("reconciling httproute")
+
+	httpRoute := model.GetGrafanaHTTPRoute(cr, scheme)
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, httpRoute, func() error {
+		httpRoute.Spec = getHTTPRouteSpec(cr, scheme)
+
+		err := v1beta1.Merge(httpRoute, cr.Spec.HTTPRoute)
+		if err != nil {
+			setInvalidMergeCondition(cr, "HTTPRoute", err)
+			return err
+		}
+
+		removeInvalidMergeCondition(cr, "HTTPRoute")
+
+		err = controllerutil.SetControllerReference(cr, httpRoute, scheme)
+		if err != nil {
+			return err
+		}
+
+		model.SetInheritedLabels(httpRoute, cr.Labels)
+
+		return nil
+	})
+	if err != nil {
+		return v1beta1.OperatorStageResultFailed, err
+	}
+
+	// try to assign the admin url if PreferHTTPRoute is enabled
+	if cr.PreferHTTPRoute() {
+		adminURL, err := r.getHTTPRouteAdminURL(ctx, httpRoute)
+		if err != nil {
+			return v1beta1.OperatorStageResultInProgress, fmt.Errorf("httproute is not ready yet: %w", err)
+		}
+
+		if adminURL == "" {
+			return v1beta1.OperatorStageResultInProgress, fmt.Errorf("httproute spec is incomplete")
+		}
+
+		cr.Status.AdminURL = adminURL
+	}
+
+	return v1beta1.OperatorStageResultSuccess, nil
+}
+
+func getHTTPRouteSpec(cr *v1beta1.Grafana, scheme *runtime.Scheme) gatewayv1.HTTPRouteSpec {
+	service := model.GetGrafanaService(cr, scheme)
+	port := gatewayv1.PortNumber(GetGrafanaPort(cr)) //nolint:gosec // Port number is always valid Grafana port
+
+	return gatewayv1.HTTPRouteSpec{
+		CommonRouteSpec: gatewayv1.CommonRouteSpec{
+			ParentRefs: []gatewayv1.ParentReference{},
+		},
+		Hostnames: []gatewayv1.Hostname{},
+		Rules: []gatewayv1.HTTPRouteRule{
+			{
+				BackendRefs: []gatewayv1.HTTPBackendRef{
+					{
+						BackendRef: gatewayv1.BackendRef{
+							BackendObjectReference: gatewayv1.BackendObjectReference{
+								Name: gatewayv1.ObjectName(service.Name),
+								Port: &port,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// getHTTPRouteAdminURL returns the admin URL for accessing Grafana via HTTPRoute.
+// It performs Gateway lookup to determine the correct protocol and hostname.
+func (r *HTTPRouteReconciler) getHTTPRouteAdminURL(ctx context.Context, httpRoute *gatewayv1.HTTPRoute) (string, error) {
```

Comment:
If the functionality gets accepted, the function will need to be refactored and covered by tests. - The current complexity is too high.
Related functions need to also be covered by tests.

Reply:
Removed PreferHTTPRoute and Gateway lookup complexity to simplify initial implementation. Created #2284 to discuss this functionality separately.